src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
             and self.graph.has_node(final_index)
         ):
             raise ValueError(
-                "`final_index` must not be an existing node not in `node_list`."
+                "`final_index` cannot be an existing node that is not in `node_list`."
             )
         if final_index is None:
```

Comment:
Potential simplification together with the above:

```suggestion
if final_index is None:
    final_index = self.df.last_valid_index()
elif final_index not in node_list and self.graph.has_node(final_index):
    raise ValueError(....)
```

Reply:
Good catch!